Support preserveEntrySignatures option

AliceLanniste opened the pull request:
<!-- Thank you for contributing! -->

### Description

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->
related #3500

- [x] preserveEntrySignatures option 
- [x] preserveSignature option in EmittedChunk

AliceLanniste committed:
fix: add`preserveEntrySignature`

feat: add `preserve_signature` in EmittedChunk

feat: add `preserve_signature` in rolldown_binding

feat: add `preserve_signature` in rolldown_binding

feat: add test for `preserveEntrySignature`

AliceLanniste committed:
fix: add`preserveEntrySignature`

AliceLanniste committed:
feat: add `preserve_signature` in EmittedChunk

AliceLanniste committed:
feat: add `preserve_signature` in rolldown_binding

AliceLanniste committed:
feat: add `preserve_signature` in rolldown_binding

IWANABETHATGUY commented:
Can you clean up the pr?

hyf0 commented:
This PR seems to add the option only, not implement the feature.

AliceLanniste commented:
> This PR seems to add the option only, not implement the feature.

yeah, I will try to implement the feature in the Chunk part

IWANABETHATGUY commented:
> > This PR seems to add the option only, not implement the feature.
> 
> yeah, I will try to implement the feature in the Chunk part

It's ok to add option first, split the pr into two parts also help maintainer with reviewing.

IWANABETHATGUY commented on packages/rolldown/src/binding.d.ts:
````diff
@@ -1182,6 +1184,7 @@ export interface StaticImportEntry {
    * ```ts
    * import type { foo } from "mod";
    * import { type foo } from "mod";
+import { PreserveEntrySignaturesOption } from '../dist/types/rollup';
````
Can you remove this?

IWANABETHATGUY commented on packages/rolldown/src/options/normalized-input-options.ts:
```diff
@@ -34,4 +35,23 @@ export class NormalizedInputOptionsImpl implements NormalizedInputOptions {
+//bundlerOptions则是放入nativeBundler
+//我的想法是新增的这个属性也要加入到BundlerOptions，这样才能参与chunk的分割
+//现在在的BundlerOptions的input属性有
+// input，cwd,external,platform,shim_missing_exports
```
Can you delete those comments?

AliceLanniste commented:
> This PR seems to add the option only, not implement the feature.
hi， code generated by `preserveEntrySigature: strict` in rollup
```
//generated_dynamic.js
import { s as shared } from './generated-main.js';

globalThis.sharedDynamic = shared;

//generated_main.js
const shared = 'shared';

const unused = 'unused';
const dynamic = import('./generated-dynamic.js');

globalThis.sharedStatic = shared;

export { dynamic as d, shared as s, unused as u };
//main.js
export { d as dynamic, u as unused } from './generated-main.js';

```

the output of the current rolldown
```

import { shared } from "./lib.js";

//#region dynamic.js
globalThis.sharedDynamic = shared;

//#endregion

## lib.js


//#region lib.js
const shared = "shared";

//#endregion
export { shared };

## main.js

import { shared } from "./lib.js";

//#region main.js
const unused = "unused";
const dynamic = import("./dynamic.js");
globalThis.sharedStatic = shared;

//#endregion
export { dynamic, unused };

```

The output of the rolldown should be the same as the output of the rollup.

AliceLanniste committed:
chore: clean up

AliceLanniste committed:
Merge branch 'preserveEntry' of github.com:AliceLanniste/rolldown into preserveEntry

IWANABETHATGUY commented:
Can you clean up the pr and let's finish the options and binding, we could finish the implementation in rest pr.

AliceLanniste commented:
> Can you clean up the pr and let's finish the options and binding, we could finish the implementation in rest pr.

sorry， I forgot to push it. I have cleaned up the pr

AliceLanniste committed:
feat: complete  preserve_entry_signature option

feat: add `preserve_signature` in EmittedChunk

feat: add `preserve_signature` in rolldown_binding

chore: clean up

feat: add `preserve_signature` in rolldown_binding

chore: delete unused import

AliceLanniste committed:
Merge remote branch

AliceLanniste committed:
chore: revert

AliceLanniste committed:
feat: add tests for `preserveEntrySignatrue`

Boshen commented:
@shulaoda can you check whether we can merge this with little effort, otherwise we'll need to close this :-(

shulaoda commented:
We currently don't have the bandwidth to support this feature, so we’ve decided to close it for now.
